File: ai_router/metrics.py
"""Append-only metrics ledger for the router.

One JSON object per line in ``router-metrics.jsonl`` — no wrapping array,
additive schema (new fields never break old lines), safe to stream with jq.
Writing is best-effort and never raises: metrics must never break a routed
call that already succeeded and was already paid for.

**Tokens are recorded and dollars are not computed.** Reconciliation happens
out of band, against the vendor's own console: a repository names its own API
key per provider, so the join between these token counts and the vendor's
dollars is the key itself. Seat spend is not attributable per session and is
not estimated — ``billed_usage_unavailable`` marks the rows a seat transport
produced, and ``transport_session_id`` (the CLI conversation id) is what
``ai_router.seat_cost`` prices them by.
"""

from __future__ import annotations

import datetime
import json
import os
from pathlib import Path
from typing import Optional

_THIS_DIR = Path(__file__).parent


def _log_path(config: dict) -> Path:
    """``AI_ROUTER_METRICS_PATH`` env var > alongside the loaded config file
    > the package directory."""
    metrics_cfg = config.get("metrics", {}) or {}
    filename = metrics_cfg.get("log_filename", "router-metrics.jsonl")

    override = os.environ.get("AI_ROUTER_METRICS_PATH")
    if override:
        return Path(override)
    config_path = config.get("_config_path")
    if config_path:
        return Path(config_path).parent / filename
    return _THIS_DIR / filename


def _metrics_enabled(config: dict) -> bool:
    return bool((config.get("metrics", {}) or {}).get("enabled", True))


def record_call(
    config: dict,
    *,
    call_type: str,                    # "route" | "verify"
    task_type: str,
    model: str,                        # registry alias (or catalog id on copilot)
    provider: str,
    generation_params: dict,
    input_tokens: int,
    output_tokens: int,
    elapsed_seconds: float,
    escalated: bool,
    stop_reason: str,
    session_number: Optional[int] = None,
    requested_model_id: Optional[str] = None,
    served_model_id: Optional[str] = None,
    transport: Optional[str] = None,
    billed_usage_unavailable: Optional[bool] = None,
    transport_session_id: Optional[str] = None,
    verifier_of: Optional[str] = None,
    verdict: Optional[str] = None,
    issue_count: Optional[int] = None,
) -> None:
    """Append one record. Never raises — a write failure (disk full,
    permissions) skips silently rather than breaking the routed call."""
    if not _metrics_enabled(config):
        return

    effort = None
    thinking_on = False
    generation_params = generation_params or {}
    if provider == "anthropic":
        effort = generation_params.get("effort")
        thinking_on = bool(
            (generation_params.get("thinking") or {}).get("enabled")
        )
    elif provider == "google":
        effort = generation_params.get("thinking_level")
        budget = generation_params.get("thinking_budget")
        thinking_on = (effort is not None) or (
            budget is not None and budget != 0
        )
    elif provider == "openai":
        effort = generation_params.get("reasoning_effort")
        thinking_on = effort not in (None, "none", "minimal")

    record = {
        "timestamp": datetime.datetime.now(datetime.timezone.utc).isoformat(),
        "session_number": session_number,
        "call_type": call_type,
        "task_type": task_type,
        "model": model,
        "requested_model_id": requested_model_id,
        "served_model_id": served_model_id,
        # Tri-state: True/False only when BOTH ids are known, else null —
        # an absent id does not establish that the provider served what was
        # asked for.
        "served_model_mismatch": (
            (requested_model_id != served_model_id)
            if (requested_model_id and served_model_id)
            else None
        ),
        "provider": provider,
        "effort": effort,
        "thinking_on": thinking_on,
        "input_tokens": int(input_tokens),
        "output_tokens": int(output_tokens),
        "elapsed_seconds": round(float(elapsed_seconds), 3),
        "escalated": bool(escalated),
        "stop_reason": stop_reason,
        "transport": transport,
        "billed_usage_unavailable": billed_usage_unavailable,
        "transport_session_id": transport_session_id,
        "verifier_of": verifier_of,
        "verdict": verdict,
        "issue_count": issue_count,
    }

    try:
        path = _log_path(config)
        path.parent.mkdir(parents=True, exist_ok=True)
        with open(path, "a", encoding="utf-8") as f:
            f.write(json.dumps(record) + "\n")
    except Exception:
        pass


def load_metrics(config: dict) -> list[dict]:
    """Read every record; unparseable lines are skipped. Empty list when
    the file is missing."""
    path = _log_path(config)
    if not path.exists():
        return []
    records = []
    with open(path, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                obj = json.loads(line)
            except json.JSONDecodeError:
                continue
            if isinstance(obj, dict):
                records.append(obj)
    return records


def _seat_rows(records: list) -> list:
    """Rows a seat transport produced. Their spend is real and is not
    attributable here — ``transport_session_id`` is what prices them."""
    return [
        r for r in records if r.get("billed_usage_unavailable") is True
    ]


def print_metrics_report(config: dict) -> None:
    """Human-readable summary: token totals, per-model / per-task / per-set
    volume, and served-model mismatches."""
    records = load_metrics(config)
    if not records:
        print("(no metrics recorded yet -- router-metrics.jsonl is empty "
              "or missing)")
        return

    def _tokens(rows: list) -> int:
        return sum(
            (r.get("input_tokens", 0) or 0) + (r.get("output_tokens", 0) or 0)
            for r in rows
        )

    print("\n" + "=" * 68)
    print(f"AI ROUTER -- METRICS REPORT  ({len(records)} calls logged)")
    print("=" * 68)

    print(f"Total input tokens:   "
          f"{sum(r.get('input_tokens', 0) or 0 for r in records):,}")
    print(f"Total output tokens:  "
          f"{sum(r.get('output_tokens', 0) or 0 for r in records):,}")

    seat = _seat_rows(records)
    if seat:
        with_id = sum(1 for r in seat if r.get("transport_session_id"))
        print(f"On a seat transport:                {len(seat)} call(s) "
              "(billed_usage_unavailable)")
        print(f"                                    real spend in AI credits; "
              f"{with_id} carry the conversation id that prices them "
              "(python -m ai_router.seat_cost)")

    mismatched = [r for r in records if r.get("served_model_mismatch")]
    if mismatched:
        grouped: dict[str, int] = {}
        for r in mismatched:
            key = f"{r.get('requested_model_id')} -> {r.get('served_model_id')}"
            grouped[key] = grouped.get(key, 0) + 1
        print("\n--- Requested vs served model ---")
        print("  A dated-snapshot pin is routine; a change of model FAMILY "
              "is a different model answering.")
        for key, count in sorted(grouped.items(), key=lambda kv: -kv[1]):
            print(f"      {count:>5}x  {key}")

    print("\n--- By model ---")
    by_model: dict[str, dict] = {}
    for r in records:
        m = r.get("model", "?")
        slot = by_model.setdefault(
            m, {"records": [], "escalated": 0, "provider": r.get("provider", "?")}
        )
        slot["records"].append(r)
        if r.get("escalated"):
            slot["escalated"] += 1
    print(f"  {'model':<24} {'provider':<11} {'calls':>6} {'tokens':>12} {'esc%':>6}")
    for m, s in sorted(
        by_model.items(), key=lambda kv: -_tokens(kv[1]["records"])
    ):
        calls = len(s["records"])
        esc_pct = (100.0 * s["escalated"] / calls) if calls else 0
        print(f"  {m:<24} {s['provider']:<11} {calls:>6} "
              f"{_tokens(s['records']):>12,} {esc_pct:>5.1f}%")

    print("\n--- By task type ---")
    by_task: dict[str, list] = {}
    for r in records:
        by_task.setdefault(r.get("task_type", "?"), []).append(r)
    print(f"  {'task_type':<24} {'calls':>6} {'tokens':>12}")
    for t, rows in sorted(by_task.items(), key=lambda kv: -_tokens(kv[1])):
        print(f"  {t:<24} {len(rows):>6} {_tokens(rows):>12,}")

    sessions: dict[int, list] = {}
    for r in records:
        number = r.get("session_number")
        if isinstance(number, int):
            sessions.setdefault(number, []).append(r)
    if sessions:
        print("\n--- By session ---")
        print(f"  {'session':<40} {'calls':>6} {'tokens':>12}")
        for number, rows in sorted(sessions.items()):
            label = f"session {number}"
            print(f"  {label:<40} {len(rows):>6} {_tokens(rows):>12,}")

    print("=" * 68 + "\n")


def main() -> int:
    from .config import load_config

    print_metrics_report(load_config())
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/router/src/metrics.ts
// Append-only metrics ledger for the router.
//
// One JSON object per line in `router-metrics.jsonl` -- no wrapping array,
// additive schema (new fields never break old lines), safe to stream with jq.
// Writing is best-effort and never raises: metrics must never break a routed
// call that already succeeded and was already paid for.
//
// **Tokens are recorded and dollars are not computed.** Reconciliation happens
// out of band, against the vendor's own console: a repository names its own API
// key per provider, so the join between these token counts and the vendor's
// dollars is the key itself. Seat spend is not attributable per session and is
// not estimated -- `billed_usage_unavailable` marks the rows a seat transport
// produced, and `transport_session_id` (the CLI conversation id) is what
// `seat_cost` prices them by.

import { appendFileSync, existsSync, mkdirSync } from "node:fs";
import { dirname, resolve } from "node:path";

import { isRecord, truthy, type RouterConfig } from "./config.ts";
import { platformNewlines } from "./journal.ts";
import { ASSET_DIR } from "./paths.ts";
import { PythonFloat, dumps } from "./pythonJson.ts";
import { readText } from "./textfile.ts";
import { writeOut } from "./output.ts";

export const METRICS_PATH_ENV_VAR = "AI_ROUTER_METRICS_PATH";
const DEFAULT_LOG_FILENAME = "router-metrics.jsonl";

function record(value: unknown): Record<string, unknown> {
  return isRecord(value) ? value : {};
}

/**
 * `AI_ROUTER_METRICS_PATH` env var > alongside the loaded config file > the
 * package directory.
 */
function logPath(config: RouterConfig): string {
  const metricsConfig = record(config["metrics"]);
  const filename =
    typeof metricsConfig["log_filename"] === "string"
      ? metricsConfig["log_filename"]
      : DEFAULT_LOG_FILENAME;

  const override = process.env[METRICS_PATH_ENV_VAR];
  if (override) return override;
  // `resolve`, not `join`: Python's `Path(dir) / name` RETURNS `name` when
  // it is absolute, and `join` would paste one path onto the other. The
  // declared filename is a config value, so it can be either.
  const configPath = config["_config_path"];
  if (typeof configPath === "string" && configPath) {
    return resolve(dirname(configPath), filename);
  }
  return resolve(ASSET_DIR, filename);
}

function metricsEnabled(config: RouterConfig): boolean {
  const block = record(config["metrics"]);
  return "enabled" in block ? Boolean(block["enabled"]) : true;
}

export interface CallRecord {
  /** "route" | "verify" */
  readonly callType: string;
  readonly taskType: string;
  /** Registry alias, or the catalog id on a seat. */
  readonly model: string;
  readonly provider: string;
  readonly generationParams: Record<string, unknown> | null;
  readonly inputTokens: number;
  readonly outputTokens: number;
  readonly elapsedSeconds: number;
  readonly escalated: boolean;
  readonly stopReason: string;
  readonly sessionNumber?: number | null;
  readonly requestedModelId?: string | null;
  readonly servedModelId?: string | null;
  readonly transport?: string | null;
  readonly billedUsageUnavailable?: boolean | null;
  readonly transportSessionId?: string | null;
  readonly verifierOf?: string | null;
  readonly verdict?: string | null;
  readonly issueCount?: number | null;
}

/**
 * The row one call becomes, decided from the call alone.
 *
 * Separate from the append so the shape both routers write -- the effort and
 * thinking fields each provider spells differently, the tri-state
 * served-model flag, the float that stays a float -- is a judgement over
 * facts rather than something only a written file can show.
 */
export function metricsRow(
  call: CallRecord,
  now: Date = new Date(),
): Record<string, unknown> {
  const params = call.generationParams ?? {};
  let effort: unknown = null;
  let thinkingOn = false;
  if (call.provider === "anthropic") {
    effort = params["effort"] ?? null;
    thinkingOn = Boolean(record(params["thinking"])["enabled"]);
  } else if (call.provider === "google") {
    effort = params["thinking_level"] ?? null;
    const budget = params["thinking_budget"];
    thinkingOn =
      effort !== null ||
      (budget !== undefined && budget !== null && budget !== 0);
  } else if (call.provider === "openai") {
    effort = params["reasoning_effort"] ?? null;
    thinkingOn = !(effort === null || effort === "none" || effort === "minimal");
  }

  const requested = call.requestedModelId ?? null;
  const served = call.servedModelId ?? null;
  const row: Array<readonly [string, unknown]> = [
    ["timestamp", pythonUtcNow(now)],
    ["session_number", call.sessionNumber ?? null],
    ["call_type", call.callType],
    ["task_type", call.taskType],
    ["model", call.model],
    ["requested_model_id", requested],
    ["served_model_id", served],
    // Tri-state: true/false only when BOTH ids are known, else null -- an
    // absent id does not establish that the provider served what was asked
    // for.
    ["served_model_mismatch", requested && served ? requested !== served : null],
    ["provider", call.provider],
    ["effort", effort ?? null],
    ["thinking_on", thinkingOn],
    ["input_tokens", Math.trunc(call.inputTokens)],
    ["output_tokens", Math.trunc(call.outputTokens)],
    ["elapsed_seconds", new PythonFloat(roundHalfEven(call.elapsedSeconds, 3))],
    ["escalated", Boolean(call.escalated)],
    ["stop_reason", call.stopReason],
    ["transport", call.transport ?? null],
    ["billed_usage_unavailable", call.billedUsageUnavailable ?? null],
    ["transport_session_id", call.transportSessionId ?? null],
    ["verifier_of", call.verifierOf ?? null],
    ["verdict", call.verdict ?? null],
    ["issue_count", call.issueCount ?? null],
  ];
  return Object.fromEntries(row);
}

/**
 * Append one record. Never throws -- a write failure (disk full, permissions)
 * skips silently rather than breaking the routed call.
 */
export function recordCall(config: RouterConfig, call: CallRecord): void {
  if (!metricsEnabled(config)) return;
  try {
    const path = logPath(config);
    mkdirSync(dirname(path), { recursive: true });
    appendFileSync(path, platformNewlines(dumps(metricsRow(call)) + "\n"), {
      encoding: "utf8",
    });
  } catch {
    // Best-effort by contract; see the module header.
  }
}

/**
 * Read every record; unparseable lines are skipped. Empty list when the file
 * is missing.
 */
export function loadMetrics(config: RouterConfig): Array<Record<string, unknown>> {
  const path = logPath(config);
  if (!existsSync(path)) return [];
  const records: Array<Record<string, unknown>> = [];
  for (const raw of readText(path).split("\n")) {
    const line = raw.trim();
    if (!line) continue;
    let parsed: unknown;
    try {
      parsed = JSON.parse(line);
    } catch {
      continue;
    }
    if (isRecord(parsed)) records.push(parsed);
  }
  return records;
}

/**
 * Rows a seat transport produced. Their spend is real and is not attributable
 * here -- `transport_session_id` is what prices them.
 */
function seatRows(
  records: ReadonlyArray<Record<string, unknown>>,
): Array<Record<string, unknown>> {
  return records.filter((row) => row["billed_usage_unavailable"] === true);
}

function tokensOf(rows: ReadonlyArray<Record<string, unknown>>): number {
  let total = 0;
  for (const row of rows) {
    total += numberOf(row["input_tokens"]) + numberOf(row["output_tokens"]);
  }
  return total;
}

function numberOf(value: unknown): number {
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}

function stringOf(value: unknown, fallback: string): string {
  return value === undefined || value === null ? fallback : String(value);
}

/**
 * Human-readable summary: token totals, per-model / per-task / per-set volume,
 * and served-model mismatches.
 */
export function printMetricsReport(
  config: RouterConfig,
  write: (text: string) => void = writeOut,
): void {
  write(renderMetricsReport(loadMetrics(config)));
}

/** The report's whole text, over rows already read. */
export function renderMetricsReport(
  records: ReadonlyArray<Record<string, unknown>>,
): string {
  let out = "";
  const print = (line = ""): void => {
    out += line + "\n";
  };
  if (records.length === 0) {
    print(
      "(no metrics recorded yet -- router-metrics.jsonl is empty " + "or missing)",
    );
    return out;
  }

  const rule = "=".repeat(68);
  print("\n" + rule);
  print(`AI ROUTER -- METRICS REPORT  (${records.length} calls logged)`);
  print(rule);

  print(
    "Total input tokens:   " +
      comma(records.reduce((sum, row) => sum + numberOf(row["input_tokens"]), 0)),
  );
  print(
    "Total output tokens:  " +
      comma(records.reduce((sum, row) => sum + numberOf(row["output_tokens"]), 0)),
  );

  const seat = seatRows(records);
  if (seat.length > 0) {
    const withId = seat.filter((row) => truthy(row["transport_session_id"])).length;
    print(
      `On a seat transport:                ${seat.length} call(s) ` +
        "(billed_usage_unavailable)",
    );
    print(
      "                                    real spend in AI credits; " +
        `${withId} carry the conversation id that prices them ` +
        "(dabbler seat-cost)",
    );
  }

  const mismatched = records.filter((row) => truthy(row["served_model_mismatch"]));
  if (mismatched.length > 0) {
    const grouped = new Map<string, number>();
    for (const row of mismatched) {
      const key = `${stringOf(row["requested_model_id"], "None")} -> ${stringOf(
        row["served_model_id"],
        "None",
      )}`;
      grouped.set(key, (grouped.get(key) ?? 0) + 1);
    }
    print("\n--- Requested vs served model ---");
    print(
      "  A dated-snapshot pin is routine; a change of model FAMILY " +
        "is a different model answering.",
    );
    for (const [key, count] of sortedBy(
      [...grouped.entries()],
      ([, count]) => -count,
    )) {
      print(`      ${pad(String(count), 5)}x  ${key}`);
    }
  }

  print("\n--- By model ---");
  interface ModelSlot {
    records: Array<Record<string, unknown>>;
    escalated: number;
    provider: string;
  }
  const byModel = new Map<string, ModelSlot>();
  for (const row of records) {
    const name = stringOf(row["model"], "?");
    let slot = byModel.get(name);
    if (!slot) {
      slot = { records: [], escalated: 0, provider: stringOf(row["provider"], "?") };
      byModel.set(name, slot);
    }
    slot.records.push(row);
    if (truthy(row["escalated"])) slot.escalated += 1;
  }
  print(
    `  ${padRight("model", 24)} ${padRight("provider", 11)} ${pad("calls", 6)} ` +
      `${pad("tokens", 12)} ${pad("esc%", 6)}`,
  );
  for (const [name, slot] of sortedBy(
    [...byModel.entries()],
    ([, slot]) => -tokensOf(slot.records),
  )) {
    const calls = slot.records.length;
    const escalationPercent = calls ? (100.0 * slot.escalated) / calls : 0;
    print(
      `  ${padRight(name, 24)} ${padRight(slot.provider, 11)} ${pad(String(calls), 6)} ` +
        `${pad(comma(tokensOf(slot.records)), 12)} ${pad(
          escalationPercent.toFixed(1),
          5,
        )}%`,
    );
  }

  print("\n--- By task type ---");
  const byTask = new Map<string, Array<Record<string, unknown>>>();
  for (const row of records) {
    const key = stringOf(row["task_type"], "?");
    const rows = byTask.get(key);
    if (rows) rows.push(row);
    else byTask.set(key, [row]);
  }
  print(`  ${padRight("task_type", 24)} ${pad("calls", 6)} ${pad("tokens", 12)}`);
  for (const [task, rows] of sortedBy(
    [...byTask.entries()],
    ([, rows]) => -tokensOf(rows),
  )) {
    print(
      `  ${padRight(task, 24)} ${pad(String(rows.length), 6)} ` +
        `${pad(comma(tokensOf(rows)), 12)}`,
    );
  }

  const sessions = new Map<number, Array<Record<string, unknown>>>();
  for (const row of records) {
    const number = row["session_number"];
    // Python's `isinstance(n, int)` excludes a float and a bool alike.
    if (typeof number !== "number" || !Number.isInteger(number)) continue;
    const rows = sessions.get(number);
    if (rows) rows.push(row);
    else sessions.set(number, [row]);
  }
  if (sessions.size > 0) {
    print("\n--- By session ---");
    print(`  ${padRight("session", 40)} ${pad("calls", 6)} ${pad("tokens", 12)}`);
    for (const [number, rows] of [...sessions.entries()].sort(
      (a, b) => a[0] - b[0],
    )) {
      print(
        `  ${padRight(`session ${number}`, 40)} ${pad(String(rows.length), 6)} ` +
          `${pad(comma(tokensOf(rows)), 12)}`,
      );
    }
  }

  print(rule + "\n");
  return out;
}

/** `dabbler metrics`. */
export function main(config: RouterConfig): number {
  printMetricsReport(config);
  return 0;
}

// --- Rendering, as Python renders it -----------------------------------------

function comma(value: number): string {
  return String(value).replace(/\B(?=(\d{3})+(?!\d))/g, ",");
}

function pad(text: string, width: number): string {
  return text.padStart(width);
}

function padRight(text: string, width: number): string {
  return text.padEnd(width);
}

/** `sorted(items, key=...)`: stable, ascending, on a numeric key. */
function sortedBy<T>(items: readonly T[], key: (item: T) => number): T[] {
  return [...items].sort((a, b) => key(a) - key(b));
}

/** UTC now as `datetime.now(timezone.utc).isoformat()` spells it. */
function pythonUtcNow(now: Date = new Date()): string {
  const iso = now.toISOString();
  // JS carries milliseconds; Python's isoformat carries microseconds.
  return iso.replace(/\.(\d{3})Z$/, ".$1000+00:00");
}

/** Python's `round(x, 3)`: half-way values go to the even neighbour. */
export function roundHalfEven(value: number, digits: number): number {
  if (!Number.isFinite(value)) return value;
  const factor = 10 ** digits;
  const scaled = value * factor;
  const floor = Math.floor(scaled);
  const remainder = scaled - floor;
  let rounded: number;
  if (remainder > 0.5) rounded = floor + 1;
  else if (remainder < 0.5) rounded = floor;
  else rounded = floor % 2 === 0 ? floor : floor + 1;
  return rounded / factor;
}
